### PR TITLE
Chore(deps)  upgrade lint deps

### LINKED ISSUE #5077: Update ESLint + ESLint plugins
After #5007 is done and ESLint is enabled again we should update ESLint and all it's plugins.

TODO:

- [ ] Wait for #5007 
- [ ] Update ESLint
- [ ] Update Plugins
- [ ] Make changes to code to pass linting if needed.

### PR DESCRIPTION
**Pull request by jmetev1**
## Description
close #5077 

### Double check

- [ ] I have tested my parser changes locally with `poetry run test_parser "zone_key"`
- [x] I have run `pnpx prettier --write .` and `poetry run format` to format my changes.

### CODE REVIEW

File: web/package.json [hunk 1]
```diff
@@ -98,7 +98,6 @@
   },
   "devDependencies": {
     "@mdx-js/react": "^2.1.5",
-    "@microsoft/eslint-formatter-sarif": "^3.0.0",
```
Comment (jmetev1) [inline web/package.json:101]: not used

Comment (VIKTORVAV99) [reply web/package.json:101]: It is used in the CI.
Please re-install it.

Comment (jmetev1) [reply web/package.json:101]: done

**Comment by jmetev1** [comment]
lint and tsc now pass

File: web/src/features/charts/elements/AreaGraphLayers.tsx [hunk 2]
```diff
@@ -75,6 +75,7 @@ function AreaGraphLayers({
         // Therefore, we copy all datapoints and make sure
         // both a start and an end are present to ensure
         // proper display of missing points
+        // eslint-disable-next-line unicorn/no-useless-spread
         const datapoints = [
```
Comment (jmetev1) [inline web/src/features/charts/elements/AreaGraphLayers.tsx:79]: if you let eslint make changes it wants here, then you'll get warning when using datapoints below. i confess i don't fully understand it so i left it alone.

File: web/package.json [hunk 3]
```diff
@@ -138,28 +138,28 @@
     "@types/topojson": "^3.2.3",
     "@types/topojson-client": "^3.1.1",
     "@types/topojson-server": "^3.0.1",
-    "@typescript-eslint/eslint-plugin": "5.41.0",
-    "@typescript-eslint/parser": "5.41.0",
-    "@vitejs/plugin-react": "2.1.0",
+    "@typescript-eslint/eslint-plugin": "^5.59.9",
+    "@typescript-eslint/parser": "^5.59.9",
+    "@vitejs/plugin-react": "4.0.0",
```
Comment (VIKTORVAV99) [inline web/package.json:143]: This don't seem related to ESlint?

File: web/package.json [hunk 4]
```diff
@@ -173,7 +173,7 @@
     "storybook": "^7.0.0-alpha.47",
     "tailwindcss": "3.1.8",
     "ts-node": "^10.9.1",
-    "typescript": "4.8.4",
+    "typescript": "5.0.4",
```
Comment (VIKTORVAV99) [inline web/package.json:176]: Why is typescript updated? This is also unrelated to ESlint and should be made in a seperate PR if needed.

File: web/.eslintrc.json [hunk 5]
```diff
@@ -144,7 +144,9 @@
     ],
     "unicorn/no-nested-ternary": ["error"],
     "unicorn/no-useless-undefined": "off",
-    "unicorn/no-null": "off"
+    "unicorn/no-null": "off",
+    // typescript doesn't recognize at syntax yet
+    "unicorn/prefer-at": 0
```
Comment (VIKTORVAV99) [inline web/.eslintrc.json:149]: This was added to TypeScript 4.6 and we are already using it in several places.
See https://devblogs.microsoft.com/typescript/announcing-typescript-4-6/#target-es2022

File: web/src/features/charts/hooks/useBreakdownChartData.ts [hunk 6]
```diff
@@ -71,22 +71,14 @@ export default function useBreakdownChartData() {
...
       const isStorage = mode.includes('storage');
 
-      if (isStorage) {
-        entry.layerData[mode] = getStorageValue(
-          mode as ElectricityStorageType,
-          value,
-          valueFactor,
-          displayByEmissions
-        );
-        // TODO: handle storage
```
Comment (VIKTORVAV99) [inline web/src/features/charts/hooks/useBreakdownChartData.ts:81]: Please add back the todo comment.

File: web/src/features/charts/elements/AreaGraphLayers.tsx (hunk 2 above)
Comment (VIKTORVAV99) [reply web/src/features/charts/elements/AreaGraphLayers.tsx:79]: It is likley cause we are calling the flat function on the entire array which is making the spread redundant, but please do not add new eslint-disable comments.

**Review by VIKTORVAV99 — CHANGES_REQUESTED** [review]
There are some changes to typescript and vite plugins that are unrelated to ESLint and should be addressed in a seperate PR.

File: web/src/features/charts/hooks/useBreakdownChartData.ts (hunk 6 above)
Comment (jmetev1) [reply web/src/features/charts/hooks/useBreakdownChartData.ts:81]: fixed

File: web/src/features/charts/elements/AreaGraphLayers.tsx (hunk 2 above)
Comment (jmetev1) [reply web/src/features/charts/elements/AreaGraphLayers.tsx]: ok do you prefer this?

File: web/package.json (hunk 3 above)
Comment (jmetev1) [reply web/package.json]: fixed

File: web/package.json (hunk 4 above)
Comment (jmetev1) [reply web/package.json]: fixed

File: web/src/features/charts/elements/AreaGraphLayers.tsx [hunk 7]
```diff
@@ -75,8 +75,8 @@ function AreaGraphLayers({
         // Therefore, we copy all datapoints and make sure
         // both a start and an end are present to ensure
         // proper display of missing points
-        const datapoints = [
-          ...layer.datapoints.map((d: { data: AreaGraphElement }) => [
+        const datapoints: any[] = layer.datapoints.flatMap(
```
Comment (jmetev1) [inline web/src/features/charts/elements/AreaGraphLayers.tsx:78]: you like the explicity any better?

File: web/src/features/time/TimeController.tsx [hunk 8]
```diff
@@ -24,13 +24,13 @@ export default function TimeController({ className }: { className?: string }) {
...
   useEffect(() => {
-    if (datetimes) {
+    if (datetimes && datetimes.at(-1)) {
       // This value is stored in state to avoid flickering when switching between time averages
       // as this effect means index will be one render behind if using datetimes directly
       setNumberOfEntries(datetimes.length - 1);
       // Reset the selected datetime when data changes
       setSelectedDatetime({
-        datetimeString: dateToDatetimeString(datetimes[datetimes.length - 1]),
+        datetimeString: dateToDatetimeString(datetimes.at(-1) as Date),
```
Comment (jmetev1) [inline web/src/features/time/TimeController.tsx:33]: the behavior of .at is kind of surprising to me. like why is newWay `date | undefined` but oldWay is `date` ?  

image
image

Comment (VIKTORVAV99) [reply web/src/features/time/TimeController.tsx:33]: That looks like an error because they both evaluate to undefined if called on a zero length array.

As far as I know it should be exactly the same as calling `length - 1` so please do use `.at()`

See this for reference:
image

File: web/src/features/charts/elements/AreaGraphLayers.tsx (hunk 7 above)
Comment (VIKTORVAV99) [reply web/src/features/charts/elements/AreaGraphLayers.tsx:78]: Should work but I have to do some testing, I won't have time to do that today though.

Comment (VIKTORVAV99) [reply web/src/features/charts/elements/AreaGraphLayers.tsx:78]: If we remove the any[] cast here I think we can merge this.

**Review by VIKTORVAV99 — CHANGES_REQUESTED** [review]
Just a final check, once these are addressed I think we can merge this.

File: web/src/features/charts/elements/AreaGraphLayers.tsx (hunk 7 above)
Comment (jmetev1) [reply web/src/features/charts/elements/AreaGraphLayers.tsx:78]: Right now the any[] cast gets a warning. which i thought was ok since there are lots of those warnings. but if you remove it then you get an error below. i tried to type it better but i don't really understand what's happening here, even after reading the comment.

File: web/src/features/time/TimeController.tsx [hunk 9]
```diff
@@ -24,13 +24,13 @@ export default function TimeController({ className }: { className?: string }) {
   );
 
   useEffect(() => {
-    if (datetimes) {
+    if (datetimes && datetimes.at(-1)) {
```
[the remark this answers is not in the document]
Comment (jmetev1) [reply web/src/features/time/TimeController.tsx]: fixed

File: web/src/features/charts/elements/AreaGraphLayers.tsx (hunk 7 above)
Comment (VIKTORVAV99) [reply web/src/features/charts/elements/AreaGraphLayers.tsx:78]: It's a pretty complex type as it represent all the data for the graphs. Let's stick with the any cast for now. (Even though the CI will complain 😅)

Comment (VIKTORVAV99) [reply web/src/features/charts/elements/AreaGraphLayers.tsx:78]: Just ran it again locally and I can see no errors? Where are you seeing the error?

Comment (jmetev1) [reply web/src/features/charts/elements/AreaGraphLayers.tsx:78]: if i take the any cast out i see this: 
image

Comment (VIKTORVAV99) [reply web/src/features/charts/elements/AreaGraphLayers.tsx:78]: Let's keep it like this then, and hopefully we can get a better typing in the future.

**Comment by VIKTORVAV99** [comment]
I'm going to merge this now, thanks for updating the dependencies!
